src/shared-with-pluot-core/script-message-channels/ReactNativeMessageChannel.js

```diff
@@ -0,0 +1,87 @@
+import ScriptMessageChannel from "./ScriptMessageChannel";
+import { EventEmitter } from "events";
+
+global.callMachineToDailyJsEmitter = new EventEmitter();
+global.dailyJsToCallMachineEmitter = new EventEmitter();
+
+/**
+ * A two-way message channel between daily-js and the call machine (pluot-core),
+ * when running in a React Native context.
+ */
+export default class ReactNativeMessageChannel extends ScriptMessageChannel {
+  constructor() {
+    super();
+    this._wrappedListeners = {}; // Mapping between listeners and wrapped listeners
+    this._messageCallbacks = {};
+  }
+
+  addListenerForMessagesFromCallMachine(listener, callFrameId, thisValue) {
+    this._addListener(
+      listener,
+      global.callMachineToDailyJsEmitter,
+      thisValue,
+      "received call machine message"
+    );
+  }
+
+  addListenerForMessagesFromDailyJs(listener, callFrameId, thisValue) {
+    this._addListener(
+      listener,
+      global.dailyJsToCallMachineEmitter,
+      thisValue,
+      "received daily-js message"
+    );
+  }
+
+  sendMessageToCallMachine(message, callback) {
+    this._sendMessage(
+      message,
+      global.dailyJsToCallMachineEmitter,
+      "sending message to call machine",
+      callback
+    );
+  }
+
+  sendMessageToDailyJs(message) {
+    this._sendMessage(
+      message,
+      global.callMachineToDailyJsEmitter,
+      "sending message to daily-js"
+    );
+  }
+
+  removeListener(listener) {
+    const wrappedListener = this._wrappedListeners[listener];
+    if (wrappedListener) {
+      window.removeEventListener('message', wrappedListener);
+      delete this._wrappedListeners[listener]; 
+    }
+  }
+
+  _addListener(listener, messageEmitter, thisValue, logMessage) {
+    const wrappedListener = (evt) => {
+      const msg = evt.data;
+      // console.log(`[ReactNativeMessageChannel] ${logMessage}`, msg);
+      if (msg.callbackStamp && this._messageCallbacks[msg.callbackStamp]) {
+        // console.log('[ReactNativeMessageChannel] handling message as callback', msg);
+        const callbackStamp = msg.callbackStamp; // Storing here since the callback could delete msg.callbackStamp
+        this._messageCallbacks[callbackStamp].call(thisValue, msg);
+        delete this._messageCallbacks[callbackStamp];
```

Comment:
Is there any chance the callback could throw an error, and prevent the next line (the `delete`) from being executed?

Reply:
Oh hm, I suppose so...if this was a problem, though, it would've existed prior to this change. I didn't want to revisit this logic too much here. Here's what it looked like before:

```js
if (msg.callbackStamp && this._messageCallbacks[msg.callbackStamp]) {
  this._messageCallbacks[msg.callbackStamp].call(this, msg);
  delete this._messageCallbacks[msg.callbackStamp];
}
```

(Actually I did revisit it a bit. There was a subtle bug before in the line `delete this._messageCallbacks[msg.callbackStamp];` because it references `msg.callbackStamp`, which is usually deleted in the callback itself. So I believe we weren't ever cleaning up this callback map).